### PR TITLE
builtins: Audit bytes arguments

### PR DESCRIPTION
**Pull request by JelleZijlstra**
As a followup from https://github.com/python/typeshed/pull/7589#issuecomment-1088046438,
I audited all occurrences of bytes in builtins.pyi by reading the corresponding C code
on CPython main.

Most use the C buffer protocol, so _typeshed.ReadableBuffer is the right type. A few
check specifically for bytes and bytearray.

### CODE REVIEW

File: stdlib/builtins.pyi [hunk 1]
```diff
@@ -200,7 +200,7 @@ _NegativeInteger = Literal[-1, -2, -3, -4, -5, -6, -7, -8, -9, -10, -11, -12, -1
 
 class int:
     @overload
-    def __new__(cls: type[Self], __x: str | bytes | SupportsInt | SupportsIndex | SupportsTrunc = ...) -> Self: ...
+    def __new__(cls: type[Self], __x: str | ReadableBuffer | SupportsInt | SupportsIndex | SupportsTrunc = ...) -> Self: ...
     @overload
     def __new__(cls: type[Self], __x: str | bytes | bytearray, base: SupportsIndex) -> Self: ...
```
Comment (JelleZijlstra) [inline stdlib/builtins.pyi:205]:
```
>>> int(memoryview(b"0xdeadbeef"), 16)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: int() can't convert non-string with explicit base
>>> int(memoryview(b"123"))
123
```

Showing that the first overload accepts buffers but the second doesn't.

File: stdlib/builtins.pyi [hunk 2]
```diff
@@ -223,7 +223,7 @@ class int:
     @classmethod
     def from_bytes(
         cls: type[Self],
-        bytes: Iterable[SupportsIndex] | SupportsBytes,  # TODO buffer object argument
+        bytes: Iterable[SupportsIndex] | SupportsBytes | ReadableBuffer,
```
Comment (JelleZijlstra) [inline stdlib/builtins.pyi:226]:
```
>>> int.from_bytes([1, 2, 3])
66051
>>> int.from_bytes(memoryview(b"123"))
3224115
```

File: stdlib/builtins.pyi [hunk 3]
```diff
@@ -480,27 +480,30 @@ class bytes(ByteString):
...
         def expandtabs(self, tabsize: SupportsIndex = ...) -> bytes: ...
     else:
         def expandtabs(self, tabsize: int = ...) -> bytes: ...
 
     def find(
-        self, __sub: bytes | SupportsIndex, __start: SupportsIndex | None = ..., __end: SupportsIndex | None = ...
+        self, __sub: ReadableBuffer | SupportsIndex, __start: SupportsIndex | None = ..., __end: SupportsIndex | None = ...
     ) -> int: ...
     if sys.version_info >= (3, 8):
         def hex(self, sep: str | bytes = ..., bytes_per_sep: SupportsIndex = ...) -> str: ...
```
Comment (JelleZijlstra) [inline stdlib/builtins.pyi:501]:
```
>>> b"xy".hex(memoryview(b"x"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: sep must be str or bytes.
>>> b"xy".hex(bytearray(b"x"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: sep must be str or bytes.
```

File: stdlib/builtins.pyi [hunk 4]
```diff
@@ -512,49 +515,52 @@ class bytes(ByteString):
...
 
     def rfind(
-        self, __sub: bytes | SupportsIndex, __start: SupportsIndex | None = ..., __end: SupportsIndex | None = ...
+        self, __sub: ReadableBuffer | SupportsIndex, __start: SupportsIndex | None = ..., __end: SupportsIndex | None = ...
     ) -> int: ...
     def rindex(
-        self, __sub: bytes | SupportsIndex, __start: SupportsIndex | None = ..., __end: SupportsIndex | None = ...
+        self, __sub: ReadableBuffer | SupportsIndex, __start: SupportsIndex | None = ..., __end: SupportsIndex | None = ...
     ) -> int: ...
     def rjust(self, __width: SupportsIndex, __fillchar: bytes = ...) -> bytes: ...
```
Comment (JelleZijlstra) [inline stdlib/builtins.pyi:534]:
```suggestion
    def rjust(self, __width: SupportsIndex, __fillchar: bytes | bytearray = ...) -> bytes: ...
```

```
>>> b"xy".rjust(5, bytearray(b"x"))
b'xxxxy'
```
(but memoryview fails)

File: stdlib/builtins.pyi [hunk 5]
```diff
@@ -512,49 +515,52 @@ class bytes(ByteString):
...
+    def maketrans(__frm: ReadableBuffer, __to: ReadableBuffer) -> bytes: ...
     def __len__(self) -> int: ...
     def __iter__(self) -> Iterator[int]: ...
     def __hash__(self) -> int: ...
     @overload
     def __getitem__(self, __i: SupportsIndex) -> int: ...
     @overload
     def __getitem__(self, __s: slice) -> bytes: ...
-    def __add__(self, __s: bytes) -> bytes: ...
+    def __add__(self, __s: ReadableBuffer) -> bytes: ...
```
Comment (JelleZijlstra) [inline stdlib/builtins.pyi:563]:
```
>>> b"x" + memoryview(b"y")
b'xy'
```

File: stdlib/builtins.pyi [hunk 6]
```diff
@@ -667,14 +679,14 @@ class bytearray(MutableSequence[int], ByteString):
     @overload
     def __setitem__(self, __s: slice, __x: Iterable[SupportsIndex] | bytes) -> None: ...
     def __delitem__(self, __i: SupportsIndex | slice) -> None: ...
-    def __add__(self, __s: bytes) -> bytearray: ...
-    def __iadd__(self: Self, __s: Iterable[int]) -> Self: ...
```
Comment (JelleZijlstra) [inline stdlib/builtins.pyi:671]: This was wrong; `ba += [1, 2, 3]` fails

File: stdlib/builtins.pyi [hunk 7]
```diff
@@ -1352,7 +1364,7 @@ def open(
     closefd: bool = ...,
     opener: _Opener | None = ...,
 ) -> IO[Any]: ...
-def ord(__c: str | bytes) -> int: ...
+def ord(__c: str | bytes | bytearray) -> int: ...
```
Comment (JelleZijlstra) [inline stdlib/builtins.pyi:1367]:
```
>>> ord(memoryview(b"x"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ord() expected string of length 1, but memoryview found
```

File: stdlib/builtins.pyi (hunk 1 above)
Comment (hauntsaninja) [reply stdlib/builtins.pyi:205]: https://github.com/bradfitz/deadbeef

File: stdlib/builtins.pyi [hunk 8]
```diff
@@ -512,49 +515,52 @@ class bytes(ByteString):
     def isspace(self) -> bool: ...
     def istitle(self) -> bool: ...
     def isupper(self) -> bool: ...
-    def join(self, __iterable_of_bytes: Iterable[ByteString | memoryview]) -> bytes: ...
-    def ljust(self, __width: SupportsIndex, __fillchar: bytes = ...) -> bytes: ...
+    def join(self, __iterable_of_bytes: Iterable[ReadableBuffer]) -> bytes: ...
+    def ljust(self, __width: SupportsIndex, __fillchar: bytes | bytearray = ...) -> bytes: ...
```
Comment (srittau) [inline stdlib/builtins.pyi:519]: Unfortunately, this will also accept `memoryview` at the moment, but having it more explicit can't hurt.

**Review by srittau — APPROVED** [review]
Thanks, I didn't double check, but the changes look reasonable.

File: stdlib/builtins.pyi (hunk 8 above)
Comment (JelleZijlstra) [reply stdlib/builtins.pyi:519]: That's a mypy bug :)

Comment (srittau) [reply stdlib/builtins.pyi:519]: It's working [as documented](https://docs.python.org/3/library/typing.html?highlight=typing%20bytearray#typing.ByteString). In the past when reviewing I've always asked people to remove `bytearray` from argument types due to that.

**Pull request merged**